### PR TITLE
Integer.parse and Decimal.parse improvements

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Converting `Integer.parse` into a builtin and making sure it can parse big values like `100!`. Adding `locale` parameter to `Locale.parse` and making sure it parses `32,5` as `32.5` double in Czech locale.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso [hunk 1]
```diff
@@ -975,13 +975,15 @@ type Integer
 
             Integer.parse "20220216"
     parse : Text -> Text -> Integer ! Parse_Error
-    parse text (radix=10) =
-        Panic.catch NumberFormatException (Long.parseLong text radix) _->
-             Error.throw (Number_Parse_Error.Error text)
+    parse text (radix=10) = Integer.parse_internal text radix
+
+    parse_internal text radix = @Builtin_Method "Integer.parse"
```
Comment (hubertp) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso]:
```suggestion
    parse_builtin text radix = @Builtin_Method "Integer.parse"
```

Comment (hubertp) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso]: Should there be a `PRIVATE` modifier explaining that users should use `parse` like in the other similar cases? See date-related builtins.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/integer/ParseIntegerNode.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      noEx1.enter();
+      try {
+        return new EnsoBigInteger(new BigInteger(t, r));
+      } catch (NumberFormatException ex2) {
+        ex = ex2;
+      }
+      noEx2.enter();
+      var errors = Context.get(this).getBuiltins().error();
+      var err = errors.makeNumberParseError(ex.getMessage());
+      return DataflowError.withoutTrace(err, this);
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/integer/ParseIntegerNode.java:36]: I recall @radeusgd not being a fan of dropping traces but I will let him comment :)

**Review by hubertp — COMMENTED** [review]
Re builtins and (lack of) support for default arguments - yes, there is a ticket https://www.pivotaltracker.com/story/show/183817602

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso]: +1 for `PRIVATE`

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso [hunk 3]
```diff
@@ -975,13 +975,15 @@ type Integer
...
 
 ## UNSTABLE
 
+
    A syntax error when parsing a double.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso]: Why another newline here?

I guess we are usually doing
```suggestion
## UNSTABLE

   A syntax error when parsing a double.
```
or
```suggestion
## UNSTABLE
   A syntax error when parsing a double.
```
 but I don't recall two empty lines there.

(We should probably standarize if there's 1 or 0 too...)

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/integer/ParseIntegerNode.java (hunk 2 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/integer/ParseIntegerNode.java:36]: Indeed I am, but currently most of these methods do _not_ include traces. So this is consistent and I'm totally fine with it - if we want to add traces we should probably modify other places too, as primarily I think we should be consistent across the lib.

File: test/Tests/src/Data/Numbers_Spec.enso [hunk 4]
```diff
@@ -220,6 +230,12 @@ spec =
             Decimal.parse "000000.0001" . should_equal 0.0001
             Decimal.parse "aaaa" . should_fail_with Number_Parse_Error.Error
 
+        Test.specify "decimal should parse hundred factorial well" <|
+            txt = hundred_factorial.to_text + ".345"
+            decimal = Decimal.parse txt
+            is_huge = decimal > (hundred_factorial / 5)
```
Comment (radeusgd) [inline test/Tests/src/Data/Numbers_Spec.enso:246]:
```suggestion
            is_huge = decimal > hundred_factorial
```

Why the `/ 5`? Seems like it should be larger than the factorial itself? Or is it due to lack of precision in decimals?

Still, seems like `0.9 * hundred_factorial` should be more than enough, vs `0.2`, no?

**Review by radeusgd — APPROVED** [review]

**Review by jdunkerley — APPROVED** [review]
As discussed be useful to be able to override the constructor name more generally on built ins

File: test/Tests/src/Data/Numbers_Spec.enso (hunk 4 above)
Comment (JaroslavTulach) [reply test/Tests/src/Data/Numbers_Spec.enso:246]: I am not sure what are the _deltas_ between `double` values at that size. Yes, probably `0.9` would work. But I just wanted to make sure something big is being computed.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso (hunk 1 above)
Comment (JaroslavTulach) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso]: Marked private at 7c5bcc600

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/NumberParseError.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import java.util.List;
+
+import org.enso.interpreter.dsl.BuiltinType;
+import org.enso.interpreter.node.expression.builtin.Builtin;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+
+@BuiltinType
+public final class NumberParseError extends Builtin {
+  @Override
+  protected final List<Cons> getDeclaredConstructors() {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/NumberParseError.java:12]: CCing @jdunkerley - this is the place where I need to change the name of declared constructor and thus I couldn't use `UniquelyConstructibleBuiltin` as it existed in `develop` branch.

**Review by hubertp — APPROVED** [review]

**Review by jdunkerley — APPROVED** [review]

File: test/Tests/src/Data/Numbers_Spec.enso (hunk 4 above)
Comment (radeusgd) [reply test/Tests/src/Data/Numbers_Spec.enso:246]: :ok:

**Pull request merged**